KeyDistribution_E91/ReferenceImplementation.qs

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+//////////////////////////////////////////////////////////////////////
+// This file contains reference solutions to all tasks.
+// The tasks themselves can be found in Tasks.qs file.
+// but feel free to look up the solution if you get stuck.
+//////////////////////////////////////////////////////////////////////
+
+namespace Quantum.Kata.KeyDistributionE91 {
+
+    open Microsoft.Quantum.Arrays;
+    open Microsoft.Quantum.Measurement;
+    open Microsoft.Quantum.Canon;
+    open Microsoft.Quantum.Intrinsic;
+    open Microsoft.Quantum.Diagnostics;
+    open Microsoft.Quantum.Convert;
+    open Microsoft.Quantum.Math;
+    open Microsoft.Quantum.Random;
+
+
+    //////////////////////////////////////////////////////////////////
+    // Part I. Preparation
+    //////////////////////////////////////////////////////////////////
+	
+    // Task 1.1. Entangled Pairs
+    operation EntangledPairs_Reference (qsAlice : Qubit[], qsBob : Qubit[]) : Unit {
+        Fact(Length(qsAlice) == Length(qsBob), "Alice and Bob should have the same number of qubits");
+
+        for i in IndexRange(qsAlice) {
+            H(qsAlice[i]);
+            CNOT(qsAlice[i], qsBob[i]);
+        }
+    }
+
+
+    //////////////////////////////////////////////////////////////////
+    // Part II. E91 Protocol
+    //////////////////////////////////////////////////////////////////
+
+    // Task 2.1 Rotate and Measure
+    operation RotateAndMeasure_Reference (q : Qubit, rotationMultiplier: Int) : Result {
+        mutable result = Zero;
+        let rotationAngle = PI() * IntAsDouble(rotationMultiplier) / 4.0;
+
+        within {
+            Ry(rotationAngle, q);
+        } apply {
+            set result = M(q);
+        }
+
+        return result;
+    }
+
+    // Task 2.2 Measure Qubit Arrays
+    operation MeasureQubitArray_Reference (qs: Qubit[], basisList: Int[]) : (Int[], Result[]) {
+        mutable results = new Result[0];
+        mutable bases = new Int[0];
+
+        for q in qs {
+            let basis = basisList[DrawRandomInt(0,2)];
+            let res = RotateAndMeasure(q, basis);
+
+            set bases += [basis];
+            set results += [res];
+        }
+
+        return (bases, results);
+
+    }
+
+    // Task 2.3 Generate the shared key
+    function GenerateSharedKey_Reference (basesAlice: Int[], basesBob: Int[], results: Result[]) : Bool[] {
+        mutable key = new Bool[0];
+
+        for (a, b, r) in Zipped3(basesAlice, basesBob, results) {
+            if a == b {
```

Comment:
This is a placeholder solution, right? There should be cases in which Alice's and Bob's bases are different but they can still use the results because the bases are compatible.